src/script_lua.c

```diff
@@ -826,13 +796,6 @@ static int luaRedisGenericCommand(lua_State *lua, int raise_error) {
     if (ldbIsEnabled())
         ldbLogRedisReply(reply);
 
-    /* Sort the output array if needed, assuming it is a non-null multi bulk
-     * reply as expected. */
-    if ((c->cmd->flags & CMD_SORT_FOR_SCRIPT) &&
```

Comment:
Do we also need to get rid of the `to-sort` flag and all related code? If we don't check `CMD_SORT_FOR_SCRIPT` anymore then we can delete it too no need to have any commands marked with it.

Reply:
I had some chats about this topic with @yossigo and realized that i think it's better to keep these flags for the sake of #9359.
i.e. we wanna provide clients and anyone else, all there is to know about redis commands, and one of the things to know is when a command has a non-deterministic reply (random content or random order).
This also means two additional changes:
* RM_CreateCommand should be able to set that flag (same as it sets the `no-monitor` flag and others)
* maybe we wanna give a better name for that flag in COMMAND command (i.e. instead of `sort_for_script`), and maybe even move it to another flags field?
* HGETALL should be set with that flag (today it has the `random` flag because the Lua code was limited and wasn't able to sort it's output.

@guybe7 FYI / please comment.
I suppose some of that work can be done in another PR, but we need this one to at least avoid deleting the flag.

Reply:
@oranagra i didn't read the full PR, please remind me why did we need to sort in the past, and now we don't anymore?

the comment says:
```
 * to-sort:     Sort command output array if called from script, so that the
 *              output is deterministic. When this flag is used (not always
 *              possible), then the "random" flag is not needed.
 ```
 but why does it matter if we're in a script or not? i would expect SUNION to behave the same as a script running SUNION and replying with its reply.
 
 anyway, my observations:
 1. SCRIPT_RANDOM_DIRTY is no longer needed
 2. "to-sort" flag is no longer needed (we will use "random" instead)
 3. "random" should be a doc-flag rather than a command flag

Reply:
@guybe7 it was needed since EVAL was propagated to the replica as is, and redis needed to make sure it'll behave there exactly the same. (e.g. if you do HVALS and only use the first 2).
it also used to forbid writes after using an random command (LIKE RANDOMKEY).

so now that redis doesn't need that anymore, we can delete it, but i think that instead of deleting it, it should be a client hint (#9876)
however, a few observations:
1. i think there's still a difference between an random command (like RANDOMKEY), and one that may just return the same results at a different order (HVALS).
2. HGETALL was incorrectly flagged because of limitations of the sorting mechanism.
3. SPOP was flagged as raondom, but i suppose it's too late (it's a write command of it's own).
4. the `sort_for_script` flag was already documented, but maybe we can "break" things and move + rename it.

bottom line, i want this PR to just delete all the logic about sorting and such, but keep the command flags, and i want #9876 to take these flags to the right place.

Reply:
yes I agree... both "random" and "sort_for_script" (maybe "random-order"..?) should be command hints, not flags

can't we drop "random" for HGATLL, SPOP, and possibly other discriminated commands? we remove it from flags, so we kinda break the output anyway...

Reply:
observation:
`$ grep -rn "CMD_RANDOM" .
./src/server.h:193:#define CMD_RANDOM (1ULL<<7)           /* "random" flag */
./src/server.c:4533:            c->flags |= CMD_RANDOM;
./src/server.c:5781:    flagcount += addReplyCommandFlag(c,cmd->flags,CMD_RANDOM, "random");
./src/module.c:797:        else if (!strcasecmp(t,"random")) flags |= CMD_RANDOM;`

I think we can re-define them: "to-sort"(renaming it as "random-order"  👍) commands would produce random outputs, while "random"(needs renaming too) commands would lead to random operations to DB. In that ways these two flags are untied and can both exist on a single command.

> bottom line, i want this PR to just delete all the logic about sorting and such, but keep the command flags, and i want #9876 to take these flags to the right place.

I did an redo commit, these flags have been re-added to codebase, maybe you missed the commit out? @oranagra

Reply:
@zhugezy i didn't miss that commit.. just wanted to state i wanna keep it that way (since the previous instruction was to re-do that after Guy's PR is merged).

p.s. i think the `random` flag doesn't mean it makes random changes in the data set (e.g. RANDOMKEY), it means the reply is random (not just an ordering issue). SPOP is a special case of a command that's both `random` and `write`.

Reply:
Re p.s.: @oranagra What I meant is not their definitions now, but that we can distinguish these commands with "random-order"/"random-update"/... further, even if we are making some small changes on them I think. And that's the meaning of **re-define**.

Reply:
closing this thread, moved the conclusions to #9876